src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,573 @@
+   */
+  std::shared_ptr<AbstractExpression> right_operand;
+
+  PredicateCondition condition;
+};
+
+/**
+ * Collected information about the predicates that need to be pulled up and other nodes that need to be adjusted.
+ */
+struct PredicatePullUpInfo {
+  /**
+   * Predicates to pull up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom.
+   */
+  std::vector<PredicateInfo> predicates;
+
+  /**
+   * Nodes that contain the predicates in the subqueries LQP.
+   */
+  std::set<std::shared_ptr<PredicateNode>> predicate_nodes;
+
+  /**
+   * Projection nodes in the sub-tree that need to be removed to make the predicate pull-up safe.
+   */
+  std::vector<std::shared_ptr<ProjectionNode>> projection_nodes;
+
+  /**
+   * Alias nodes in the sub-tree that need to be extended to make the predicate pull-up safe.
+   *
+   * The nodes are ordered top to bottom.
+   */
+  std::vector<std::shared_ptr<AliasNode>> alias_nodes;
+
+  /**
+   * Aggregates that need to be patched allow the predicates to be pulled up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom. The number notes the number of predicates above this
+   * aggregate in the LQP.
+   */
+  std::vector<std::pair<std::shared_ptr<AggregateNode>, size_t>> aggregate_nodes;
+};
+
+/**
+ * Checks whether a predicate node should be turned into a join predicate.
+ */
+std::optional<PredicateInfo> should_become_join_predicate(
+    const std::shared_ptr<PredicateNode>& predicate_node,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping, bool is_below_aggregate) {
+  // Check for the type of expression first. Note that we are not concerned with predicates of other forms using
+  // correlated parameters here. We check for parameter usages that prevent optimization later in
+  // contains_unoptimizable_correlated_parameter_usages.
+  if (predicate_node->predicate()->type != ExpressionType::Predicate) {
+    return std::nullopt;
+  }
+
+  const auto& predicate_expression = std::static_pointer_cast<AbstractPredicateExpression>(predicate_node->predicate());
+
+  // Joins only support these six binary predicates. We rely on PredicateSplitUpRule having split up ANDed chains of
+  // such predicates previously, so that we can process them separately.
+  auto cond_type = predicate_expression->predicate_condition;
+  if (cond_type != PredicateCondition::Equals && cond_type != PredicateCondition::NotEquals &&
+      cond_type != PredicateCondition::LessThan && cond_type != PredicateCondition::LessThanEquals &&
+      cond_type != PredicateCondition::GreaterThan && cond_type != PredicateCondition::GreaterThanEquals) {
+    return std::nullopt;
+  }
+
+  // We can currently only pull equals predicates above aggregate nodes. The other predicate types could be supported
+  // but require more sophisticated reformulations.
+  if (is_below_aggregate && cond_type != PredicateCondition::Equals) {
```

Comment:
We group by the attribute that gets compared in the equals predicate (so instead of summing all rows where that attribute has a specific value, we group by the attribute and build the sum for each group)